miner/src/pool/queue.rs

```diff
@@ -458,6 +458,7 @@ impl TransactionQueue {
 			let state_readiness = ready::State::new(client.clone(), stale_id, nonce_cap);
 			removed += self.pool.write().cull(Some(chunk), state_readiness);
 		}
+		self.cached_pending.write().clear();
```

Comment:
This is not strictly needed:
1. The cache will be cleared if you request for a different block number
2. The cache will be cleared if you add any new transactions succesfully
3. The old values are still valid (for given (historical) `block_number`), so for instance we can still use it for propagation.

That said, I don't think it makes any difference.

Reply:
Actually, I still think it's needed. I've seen this happening (with one sender):

1. A block is created and imported into the chain, which includes transaction w/ nonce N.
2. A transaction is sent with nonce N+1, which triggers a new block creation.
3. The new block tries to import pending transactions, but the transaction queue still includes tx w/ nonce N. As the previous block has been imported, the tx w/ nonce N cannot be imported, and so the new pending block has 0 transactions.
4. The transaction queue is culled, which leaves tx w/ nonce N + 1, which triggers a new block, but at the same height as the previous one, so the cache has already been created with 0 transaction.

Reply:
I think if we changed the _transaction pool_ [here](https://github.com/paritytech/parity-common/blob/master/transaction-pool/src/pool.rs#L626) to take into account the case `Readiness::Stale` ([see here](https://github.com/paritytech/parity-ethereum/blob/master/miner/src/pool/ready.rs#L96)) we could fix this issue. What do you think @tomusdrw ?

Reply:
@ngotchac Indeed! I was sure that we just skip `Stale` transactions in the `PendingIterator`, but it seems we don't. Definitely [here](https://github.com/paritytech/parity-common/blob/master/transaction-pool/src/pool.rs#L626) we should just check next-best in case we get `Stale`.

Reply:
cf. https://github.com/paritytech/parity-common/pull/218